File: CombinationSum/II/CombinationsSumII.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        vector<vector<int>> result;
        vector<int> subCombination;
        sort(candidates.begin(), candidates.end());
        BACK(result, subCombination, target, candidates, 0);
        return result;
    }
private:
	void BACK(vector<vector<int>>& result, vector<int>& subCombination, int target, vector<int>& candidates, int start) {
		if(target == 0) {
			result.push_back(subCombination);
			return;
		}
		for(int i = start; i < candidates.size(); i++) {
			if(candidates[i]>target)
				return;
			if(i == start || candidates[i] != candidates[i - 1]) {
				subCombination.push_back(candidates[i]);
				BACK(result, subCombination, target - candidates[i], candidates, i + 1);
				subCombination.pop_back();
			}
		}
	}
};

File: CombinationSum/II/CombinationsSumII.py
class Solution(object):
    def combinationSum2(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        res = []
        getSubSum2(candidates, target, res, [], 0)
        return res

def getSubSum2(candidates, target, res, tmp, cur):
    if target == 0:
        res += tmp,
        return res
    else:
        for i in range(cur, len(candidates)):
            if candidates[i] > target:
                return
            if i != 0 and candidates[i] == candidates[i - 1] and i > cur:
                continue
            getSubSum2(candidates, target - candidates[i], res, tmp + [candidates[i]], i + 1)